WIP: egraph-based midend: draw the rest of the owl (productionized).

cfallin opened the pull request:
This PR is a draft of an updated version of the egraph patch (and thus supersedes #4249) with the two parts already merged (multi-etors and the egraph crate proper) removed; it includes the Cranelift integration, the egraph build (CLIF to egraph) and elaboration (egraph to CLIF) algorithms, and rule application engine, as well as a set of rewrite rules that replaces the existing mid-end optimizations.

It still needs a bit more productionizing:
- removal of recursion in elaboration;
- removal of recursion in rule application (This one is trickier! Immediate rule application on the sub-nodes created from constructors means more than one ISLE invocation can be on the stack, in a reentrant way. My thought is to use a sort of workqueue to "unstack" it.);
- generalization of the several ad-hoc egraph analyses (loop depth, etc) into a framework.

The purpose of this draft PR is to be a place to do this work on a rebased and up-to-date basis. (Lots happened since the original egraph work branched off in May, including incremental compilation and a good number of smaller changes.)

While patch-wrangling this week, I tried pulling this apart into smaller pieces, but the remaining bits are pretty cyclically entangled, and/or some of the intermediate points that might make sense (e.g. egraph build and elaboration without rule application) require re-synthesizing some scaffolding that would then disappear in the final state, so that seems a bit counterproductive. Once we have a polished state I can try pulling it apart into separate logical commits at least.

cfallin commented:
@jameysharp I think this is more-or-less ready for review now!

(In fact, I just spent a few hours doing the latest rebase, and these are fairly painful because this PR does a lot of code-motion to factor out ISLE commonalities for mid-end and backends, so I'd really prefer to try to get this merged before much more changes to avoid another rebase hell :-) )

I've updated the implementations of the various rewriting and elaboration algorithms to avoid explicit recursion, and I've factored out a generic "analysis" mechanism for the egraph crate to own; these were the main cleanups I wanted/needd to do before the prototype would be considered production-ready, I think. Regarding the rule-rewriting reentrancy issue, I opted to simply cap the recursion with a static limit (max of 5 chained rewrites) for now, as (i) we'll need such a limit anyway to protect against accidentally infinitely-(co)recursive rules, and (ii) it is about an order of magnitude simpler than unwinding the nested-immediate-rewrites into some sort of "pending node queue" and splitting the id-space, and probably more efficient in the common case.

In any case, let me know what you think -- I'm eager for feedback!

cfallin commented:
@alexcrichton it looks like the verify-publish CI job is failing [here](https://github.com/bytecodealliance/wasmtime/actions/runs/3179742540/jobs/5182545720#step:6:810) because for some reason it's looking for the (not-yet-published) cranelift-egraph crate on crates.io rather than in the workspace. I copied the magic incantations for workspace-inheritance from the other intra-repo deps but I'm not sure if I'm missing something -- it's surprising that this fails if all the other builds work fine?

nivkner commented:
tested using `cranelift-tools` with the following `clif` code:
```
target x86_64 

function u0:359(i64) -> i8 system_v {

    ss0 = explicit_slot 16
    sig0 = (i64) -> i8, i8 system_v
    fn0 = colocated u0:521 sig0

    block0(v0: i64):
	v3, v4 = call fn0(v0)
	return v3
}
```
when calling
```
./target/debug/clif-util compile --set use_egraphs=false --target x86_64-unknown-linux-gnu example.clif
```
the code compiles with out error.

when setting `--set use_egraphs=true`
the following error is produced:
```
thread 'main' panicked at 'enode depends directly on multi-value result', cranelift/codegen/src/egraph/elaborate.rs:396:33
```

<details>
  <summary>backtrace</summary>
  
  ```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: cranelift_codegen::egraph::elaborate::Elaborator::process_elab_stack::{{closure}}
             at ./cranelift/codegen/src/egraph/elaborate.rs:396:33
   3: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:301:13
   4: core::option::Option<T>::map
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/option.rs:929:29
   5: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/iter/adapters/map.rs:103:9
   6: <smallvec::SmallVec<A> as core::iter::traits::collect::Extend<<A as smallvec::Array>::Item>>::extend
             at /home/user/.local/share/cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-1.8.0/src/lib.rs:1742:36
   7: <smallvec::SmallVec<A> as core::iter::traits::collect::FromIterator<<A as smallvec::Array>::Item>>::from_iter
             at /home/user/.local/share/cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-1.8.0/src/lib.rs:1727:9
   8: core::iter::traits::iterator::Iterator::collect
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/iter/traits/iterator.rs:1836:9
   9: cranelift_codegen::egraph::elaborate::Elaborator::process_elab_stack
             at ./cranelift/codegen/src/egraph/elaborate.rs:391:60
  10: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_eclass_use
             at ./cranelift/codegen/src/egraph/elaborate.rs:285:9
  11: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_block
             at ./cranelift/codegen/src/egraph/elaborate.rs:511:13
  12: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_domtree
             at ./cranelift/codegen/src/egraph/elaborate.rs:532:21
  13: cranelift_codegen::egraph::elaborate::Elaborator::elaborate
             at ./cranelift/codegen/src/egraph/elaborate.rs:583:9
  14: cranelift_codegen::egraph::FuncEGraph::elaborate
             at ./cranelift/codegen/src/egraph.rs:348:9
  15: cranelift_codegen::context::Context::optimize
             at ./cranelift/codegen/src/context.rs:203:13
  16: cranelift_codegen::context::Context::compile_stencil
             at ./cranelift/codegen/src/context.rs:149:13
  17: cranelift_codegen::context::Context::compile
             at ./cranelift/codegen/src/context.rs:224:23
  18: cranelift_codegen::context::Context::compile_and_emit
             at ./cranelift/codegen/src/context.rs:132:29
  19: clif_util::compile::handle_module
             at ./cranelift/src/compile.rs:71:33
  20: clif_util::compile::run
             at ./cranelift/src/compile.rs:46:9
  21: clif_util::main
             at ./cranelift/src/clif-util.rs:106:33
  22: core::ops::function::FnOnce::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
  ```
</details>

alexcrichton commented:
>it's surprising that this fails if all the other builds work fine? 

You'll need to edit [this list](https://github.com/bytecodealliance/wasmtime/blob/c9ff14e00bb0a90905a4f1cc2968c1e3e0417ce5/scripts/publish.rs#L18-L67) to include automatic publishing of a new crate.

cfallin commented:
> > it's surprising that this fails if all the other builds work fine?
> 
> You'll need to edit [this list](https://github.com/bytecodealliance/wasmtime/blob/c9ff14e00bb0a90905a4f1cc2968c1e3e0417ce5/scripts/publish.rs#L18-L67) to include automatic publishing of a new crate.

It’s actually already there, from the earlier PR that introduced the crate (line 28); this seems to be something new related to the workspaces change?

alexcrichton commented:
Ah I only checked the diff of this PR, the list also needs to be topologically sorted so if `cranelift-codegen` is picking up a new dep then the list needs to be reordered.

cfallin commented:
Ah, that makes sense, thanks!

cfallin commented:
> when setting use_egraphs=true the following error is produced:
>
> ```
> thread 'main' panicked at 'enode depends directly on multi-value result', cranelift/codegen/src/egraph/elaborate.rs:396:33
> ```

Fixed, thanks! This was an issue introduced in my non-recursive rewrite of the extraction (best-cost node) procedure's handling of value projections.

nivkner commented:
the following code compiles without error when setting `use_egraphs=false` but produces an error when `use_egraphs=true`:

```
target x86_64

function u0:1302(i64) -> i64 system_v {

        block0(v0: i64):
                v9 = atomic_rmw.i64 add v0, v0
                return v0
}
```

the error is
```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', /home/user/Documents/wasmtime/target/debug/build/cranelift-codegen-ae89adb74262e952/out/opcodes.rs:1979:39
```

<details>
<summary>backtrace</summary>

```
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:142:14
   2: core::panicking::panic_bounds_check
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/panicking.rs:84:5
   3: cranelift_codegen::ir::instructions::InstructionImms::with_args
             at ./target/debug/build/cranelift-codegen-ae89adb74262e952/out/opcodes.rs:1979:39
   4: cranelift_codegen::egraph::elaborate::Elaborator::add_node
             at ./cranelift/codegen/src/egraph/elaborate.rs:197:17
   5: cranelift_codegen::egraph::elaborate::Elaborator::process_elab_stack
             at ./cranelift/codegen/src/egraph/elaborate.rs:445:35
   6: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_eclass_use
             at ./cranelift/codegen/src/egraph/elaborate.rs:285:9
   7: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_block
             at ./cranelift/codegen/src/egraph/elaborate.rs:513:13
   8: cranelift_codegen::egraph::elaborate::Elaborator::elaborate_domtree
             at ./cranelift/codegen/src/egraph/elaborate.rs:534:21
   9: cranelift_codegen::egraph::elaborate::Elaborator::elaborate
             at ./cranelift/codegen/src/egraph/elaborate.rs:585:9
  10: cranelift_codegen::egraph::FuncEGraph::elaborate
             at ./cranelift/codegen/src/egraph.rs:348:9
  11: cranelift_codegen::context::Context::optimize
             at ./cranelift/codegen/src/context.rs:203:13
  12: cranelift_codegen::context::Context::compile_stencil
             at ./cranelift/codegen/src/context.rs:149:13
  13: cranelift_codegen::context::Context::compile
             at ./cranelift/codegen/src/context.rs:224:23
  14: cranelift_codegen::context::Context::compile_and_emit
             at ./cranelift/codegen/src/context.rs:132:29
  15: clif_util::compile::handle_module
             at ./cranelift/src/compile.rs:71:33
  16: clif_util::compile::run
             at ./cranelift/src/compile.rs:46:9
  17: clif_util::main
             at ./cranelift/src/clif-util.rs:106:33
  18: core::ops::function::FnOnce::call_once
             at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/core/src/ops/function.rs:248:5
```
</details>

cfallin commented:
> the following code compiles without error when setting use_egraphs=false but produces an error when use_egraphs=true:

Thanks! Fixed now -- atomic ops were incorrectly being handled as ordinary loads by the alias-analysis machinery in egraph optimization.

jameysharp commented on cranelift/codegen/src/context.rs:
```diff
@@ -159,18 +182,34 @@ impl Context {
         }
 
-        isa.compile_function(&self.func, self.want_disasm)
+        self.optimized = true;
```
I'm a little confused about this `optimized` flag.

I guess I'd like to at least see a comment about why `fn optimize` is `pub`, perhaps similar to the one on `compile_stencil`.

If `fn optimize` were not public, would it still be necessary to guard against it being called multiple times? Under what circumstances would it get called again?

Should the "already optimized" check be at the beginning of `optimize` instead of in `compile_stencil`?

I wonder, would it make sense to set `self.optimized` at the beginning of `fn optimize`? If an error occurs, I wouldn't expect the optimizer would work any better if you run it a second time. Maybe after failed optimization one might want to retry compilation without optimization?

cfallin replied:
Yes, this was an attempt to make the API "smarter" in some sense, but I agree that the don't-re-optimize behavior is in the end just unnecessary complexity. I've marked `optimize()` as "only for testing" in the doc-comment and removed the `optimized` flag.

jameysharp commented on cranelift/codegen/src/context.rs:
```diff
@@ -159,18 +182,34 @@ impl Context {
+                "About to optimize with egraph phase:\n{}",
+                self.func.display()
+            );
+            self.compute_domtree();
+            self.eliminate_unreachable_code(isa)?;
+            self.dce(isa)?;
+            self.remove_constant_phis(isa)?;
```
Why is it necessary to re-run all of these passes after three of the four have just been run? Could you let the earlier DCE pass run if `isa.flags().use_egraphs() || opt_level != OptLevel::None` and then not need any of these?

jameysharp commented on cranelift/codegen/src/ir/immediates.rs:
```diff
@@ -479,7 +479,7 @@ impl FromStr for Offset32 {
 ///
 /// All bit patterns are allowed.
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, Hash)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
 #[repr(C)]
 pub struct Ieee32(u32);
```
I'm concerned about marking `Ieee32` and `Ieee64` as `Ord`. Their `PartialOrd` implementation really is partial, since it's based on IEEE754 ordering. We might need to split these into two versions: one that respects IEEE754 except for preserving exact NaN representations, and another where ordering and equality are defined entirely by bitwise representation. I had a really hard time balancing those two competing goals recently.

cfallin replied:
Removed! This was needed for an earlier egraph implementation but no longer used, so I've removed all of the now-extraneous `Ord` derivations on IR types.

jameysharp commented on scripts/publish.rs:
```diff
@@ -88,6 +88,7 @@ const PUBLIC_CRATES: &[&str] = &[
     "cranelift-bforest",
     "cranelift-codegen-shared",
     "cranelift-codegen-meta",
+    "cranelift-egraph",
```
Now you have cranelift-egraph in this list in two places. I think that would only be necessary if there's a cyclic dependency between crates, and I think Cargo would refuse to build that... right?

cfallin replied:
Different list actually! There is the `PUBLIC_CRATES` list (the line here) and the `CRATES_TO_PUBLISH` list and they are different for reasons I don't fully grok, but the crate needs to be in both for the publish script to be happy.

jameysharp replied:
Ohhh, that makes sense.

jameysharp reviewed (commented):
I'm not done reviewing yet (GitHub tells me I'm at 26/62 files viewed) but here are some early comments.

cfallin commented on cranelift/codegen/meta/src/shared/settings.rs:
```diff
@@ -53,6 +53,17 @@ pub(crate) fn define() -> SettingGroup {
+        r#"
+            This enables an optimization phase that converts CLIF to an egraph (equivalence graph)
+            representation, performs various rewrites, and then converts it back. This can result in
+            better optimization, but at the cost of a longer compile time.
```
Yep, that bit of text is outdated! Updated to "is currently considered experimental" instead.

cfallin commented:
Thanks @jameysharp! I've addressed your comments so far and merged the latest `main` in again (including resolving overlaps in the current mid-end rules, as overlap checking is now on by default). Happy to address any other comments as you have them.

nivkner commented:
when modifying https://github.com/bjorn3/rustc_codegen_cranelift to use this PR, and running the tests of `rand_chacha` using `cargo-clif` as described in the `rustc_codegen_cranelift` repo,
without this PR all the tests pass, with this PR `test_chacha_true_values_c` fails.

jameysharp reviewed (commented):
Thanks for addressing my previous feedback! Here's another round of review. There's a lot of code to look at here. :sweat_smile:

I appreciate you renaming all the uses of "parent" to "child" in the egraph implementation. That makes my head hurt less!

cfallin commented:
> when modifying https://github.com/bjorn3/rustc_codegen_cranelift to use this PR, and running the tests of `rand_chacha` using `cargo-clif` as described in the `rustc_codegen_cranelift` repo,
> without this PR all the tests pass, with this PR `test_chacha_true_values_c` fails.

Could you see if you can minimize this to a CLIF file? Or at least, provide more details on the failure?

Reports from out-of-repo users are absolutely valuable, but are hard to act upon: generally we should strive for development of Cranelift to require only the Cranelift repo, and if failures are seen only outside of the repo, that means we need more tests in-repo. We definitely should not have to create a cg\_clif development setup to chase down bugs. If/when a bug report with sufficient detail exists, we can work through it.

Thank you for trying this out, though.

cfallin commented on cranelift/egraph/src/lib.rs:
```diff
@@ -256,6 +258,20 @@ pub trait Language: CtxEq<Self::Node, Self::Node> + CtxHash<Self::Node> {
+/// node as it is created.
+pub trait Analysis {
+    type L: Language;
+    type Value: Debug + Clone + Default;
```
> Would you mind double-checking for me that you need Analysis::Value to implement all three of Debug, Clone, and Default? I naively wouldn't have expected Default to be necessary since initial analysis results should come from Analysis::for_node.
>
> I assume both Default and Clone are because of the requirements for SecondaryMap.

Indeed, they both come from `SecondaryMap`'s constraints on its item type `V`; verified just now by attempting to remove them.

It turns out `Debug` was only needed for, well, my debugging (some `log`s I had in place); those are no longer present so I've removed that constraint. We can always put it back if we want to check in some persistent `log`s, I think.

> Then you could declare EGraph::nodes as a vector of (L::Node, A::Value), so that it's clear at the type level that if a node has been added, its analysis result exists too. And then you shouldn't need Default or Clone.

There is a nice appeal to this for sure! I actually had built the thing this way originally. I'm hesitant to go back to it chiefly because it makes the API for `for_node` more awkward: we need to provide it access to the analysis values for prior nodes (so it can depend on e.g. arguments' analysis values) and I don't like the idea of exposing internal data-structure details, like a tuple-of-node-and-value. Unfortunately we can't parameterize it on the "getter" because we can't provide a function trait-constrained type parameter to the trait method. And we can't just pass in the whole `EGraph` because the `EGraph` is parameterized on the particular `Analysis` and if I try to use `Self` there, I run into issues with unsized types (??). Basically, it becomes a type-system mess. So I'd kind of prefer to at least run with what we have for now and maybe explore more advanced solutions here later, if possible :-)

jameysharp replied:
I just had a go at doing this, and while I think it's feasible (with some `?Sized` constraints to allow unsized types if needed), the current implementation is tangled up enough in assumptions about analysis working this way that I don't want to try harder right now. So yes, let's come back to this later. :laughing:

cfallin commented on cranelift/codegen/src/ir/sourceloc.rs:
```diff
@@ -14,7 +14,7 @@ use serde::{Deserialize, Serialize};
 /// The default source location uses the all-ones bit pattern `!0`. It is used for instructions
 /// that can't be given a real source location.
-#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Nope, removed!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+//!
+//! We partition memory state into several *disjoint pieces* of
+//! "abstract state". There are a finite number of such pieces:
+//! currently, we call them "heap", "table", "vmctx", and "other".Any
```
Fixed!

cfallin commented on cranelift/codegen/src/egraph/domtree.rs:
```diff
@@ -0,0 +1,72 @@
+            let block = self.block.unwrap();
+            self.block = self.domtree.nodes[block].next;
+            Some(block)
+        }
```
Much better, thanks!

cfallin commented on cranelift/codegen/src/egraph/domtree.rs:
```diff
@@ -0,0 +1,72 @@
+            let idom = match func.layout.inst_block(idom_inst) {
+                Some(idom) => idom,
+                None => continue,
+            };
```
Yep, I've added an `expect` instead; we shouldn't have dangling instructions at this point. Thanks!

cfallin commented on cranelift/entity/src/list.rs:
```diff
@@ -464,6 +482,53 @@ impl<T: EntityRef + ReservedValue> EntityList<T> {
+
+    /// Removes duplicates from the list. Assumes the list has already
+    /// been sorted.
+    pub fn remove_dups(&mut self, pool: &mut ListPool<T>) {
```
Actually it turns out this was completely unused, as was `sort` above it (they were left over from an earlier approach that stored ... parent pointers, I think? ... in EntityLists, and used sort/dedup to do merging -- parent pointers are gone now!). Removed this part of the diff.

cfallin commented:
Just merged in `main` once more. Hopefully that's the last merge from `main` into this PR, if all goes well :-) @jameysharp I'm eagerly awaiting any more comments you have!

jameysharp commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,382 @@
+                },
+            ) => self.id_eq(value, other_value, uf) && result == other_result && ty == other_ty,
+            (
+                &Node::Pure {
+                    ref op,
+                    ref args,
+                    ref types,
+                },
```
I'm not a huge fan of the pattern of stripping off a layer of reference by sticking an `&` at the beginning of a pattern, and then putting it back with the `ref` keyword. I rather like `Node::Pure { op, args, types }` here, for instance. But I don't feel anywhere near strong enough to, y'know, fight about it or whatever.

cfallin replied:
So in isolation I'd definitely agree! However I tend to like the explicit match-the-borrow form in other cases, so we don't get `Copy` types that nonetheless have to be dereferenced. E.g. elsewhere there is a `&Node::Inst { inst, .. }` -- if we instead did `Node::Inst { inst, .. }` then we'd need `*inst` everywhere. Here I'm trying to remain consistent with that form elsewhere.

(I also think that really confusing errors can arise from this bit of automatic/ergonomic magic and I tend to fall back to pre-2018 Rust out of habit as a result, but that's another discussion :-) )

jameysharp commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,382 @@
+            }
+            &Node::Load {
+                ref op,
+                ty,
```
I appreciated the comment above that `types` are fully determined by `op` and `args`. I think it'd be nice if there's a corresponding comment here explaining why that isn't also true for Load ops. I assume it's that the type is the only thing indicating the width of the load; multiple loads with the same address can have different widths and must not be deduplicated.

cfallin replied:
Done!

jameysharp commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,382 @@
+    pub(crate) fn at_level(&self, loop_level: LoopLevel) -> Cost {
+        let loop_level = std::cmp::min(2, loop_level.level());
+        let multiplier = 1u32 << ((10 * loop_level) as u32);
+        Cost(self.0.saturating_mul(multiplier))
```
I don't see the same sort of guard here in `at_level`, for saturating at less than "infinity", that I see in the implementation for `add`. Is that okay?

cfallin replied:
Good point, fixed! I added the `finite()` helper here for this.

jameysharp reviewed (commented):
Okay, a little more review progress... I'll get there...!

jameysharp commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -29,6 +30,39 @@ pub struct LoopAnalysis {
+    }
+    /// Invalid loop level.
+    pub fn invalid() -> Self {
+        Self(0x80)
```
I'm curious, why pick 0x80 as the invalid level, instead of 0xFF?

cfallin replied:
To be honest, I have no idea -- lost to the mists of time! I've reworked this to use a sentinel of `0xff` and hopefully be a little clearer.

jameysharp reviewed (commented):
I ought to have said sooner: I believe this PR is a good idea and is overall a good implementation of the idea, so I'm reviewing the details rather than the big picture.

I'm almost done reviewing all the supporting code; I've been saving elaboration and the like for last.

cfallin commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,382 @@
+        /// because we clear the DataFlowGraph before elaboration,
+        /// this `Inst` is *not* valid to fetch any details from the
+        /// original instruction.
+        inst: Inst,
```
Updated comment, thanks!

cfallin commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,382 @@
+
+impl NodeCtx {
+    fn id_eq(&self, a: Id, b: Id, uf: &mut UnionFind) -> bool {
+        uf.find_and_update(a) == uf.find_and_update(b)
```
Good idea, done!

cfallin commented on cranelift/entity/src/list.rs:
```diff
@@ -464,6 +482,18 @@ impl<T: EntityRef + ReservedValue> EntityList<T> {
+            let elt = other.get(i, pool).unwrap();
+            *self.get_mut(offset + i, pool).unwrap() = elt;
+        }
+    }
```
Actually this was another addition from earlier work that is no longer used so I've gone ahead and deleted it. Sorry about that!

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -0,0 +1,189 @@
+                     x))
+      (subsume x))
+(rule 2 (simplify (bxor ty x x))
+      (subsume x))
```
Good catch -- `x^x` does not actually equal `x` at all, except in one rare case :-)

> Multiplication by zero

Added!

> bitwise-and by zero

Added!

jameysharp replied:
Wait, so it _is_ okay to delete instructions that might trap? Where would I find something describing whether Cranelift is allowed to do optimizations like that?

If so you could add `x|not(x)=-1` and `x&not(x)=0`, I guess.

cfallin replied:
Oh, no, we definitely don't delete any side-effecting instructions (aside from redundant loads but that's separate). I'm not sureI understand why including `x*0 == 0` or `x&0 == 0` has any effect on that?

jameysharp replied:
Okay, it's clear that these optimizations can potentially delete the only transitive data dependency on a potentially side-effecting instruction like integer division, right?

I guess the reason you're saying we don't delete those in spite of these optimizations is because they're kept even if there is no data dependency on them from any return instruction. Is that right?

In that case I think I'm satisfied. Though maybe a comment to that effect, in this source file specifically, would help somebody later.

cfallin replied:
Right, side-effecting ops are kept in the top-level "side-effect skeleton", so they are never deleted.

Do you think we need a comment specifically on these rules, or ...? The reason I ask is that strictly speaking, *any optimization at all* that removes a reference to an input could trigger this confusion. I guess a top-level note on what rules are allowed to do (rewrite pure ops arbitrarily, remove references...) could help though.

cfallin replied:
(Added a comment; feel free to suggest better clarifications if you have more ideas)

cfallin commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -71,6 +106,19 @@ impl LoopAnalysis {
+            .map(|lp| (self.loop_header(lp), lp))
+            .filter(|&(header, _)| header == block)
+            .map(|(_, lp)| lp)
+    }
```
Goodness, yes, I don't know why I wrote it that way!

cfallin commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -96,6 +144,14 @@ impl LoopAnalysis {
+        self.block_loop_map[block]
+            .expand()
+            .map(|lp| self.loops[lp].level)
+            .unwrap_or(LoopLevel(0))
```
Ah, nice, I didn't know about `map_or`! Done (and using the helper).

cfallin commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -228,6 +285,28 @@ impl LoopAnalysis {
+                            stack.push(parent);
+                        }
+                    } else {
+                        self.loops[lp].level = LoopLevel(1);
```
Yep, I agree that using `inc()` once we've defined it is cleaner here.

Re: alternative algorithm -- maybe, but the details aren't completely clear to me; happy to take a followup PR (or a suggestion here) if you can sketch it out? Although if I'm understanding what you've proposed correctly, it sounds like it may have quadratic worst-case (the algorithm here sets the level on all loops it visits as it comes back down the stack).

cfallin commented:
> To be clear this doesn't change anything if the respective setting isn't explicitly enabled? If it does change things I would like to test it with cg_clif first tomorrow.

Yes, there is no change by default; the option has to be specifically enabled. At some later point we'll want to turn it on by default but only after it's tested and mature.

jameysharp commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    fn meet_from(&mut self, other: &LastStores, loc: Inst) {
+        let meet = |a: MemoryState, b: MemoryState| -> MemoryState {
+            match (a, b) {
+                (a, b) if a == b => a,
+                (MemoryState::Top, a) | (a, MemoryState::Top) => a,
+                _ => MemoryState::BeforeInst(loc),
+            }
+        };
+
+        self.heap = meet(self.heap, other.heap);
+        self.table = meet(self.table, other.table);
+        self.vmctx = meet(self.vmctx, other.vmctx);
+        self.other = meet(self.other, other.other);
+    }
```
If you get rid of `MemoryState::Top` it becomes pretty straightforward to have `meet_from` return whether any of the states changed, which could simplify `compute_block_input_states`:

```suggestion
    fn update(&mut self, other: &LastStores, loc: Inst) -> bool {
        let meet = |a: &mut MemoryState, b: MemoryState| -> MemoryState {
            if *a == b {
                false
            } else {
                *a = MemoryState::BeforeInst(loc);
                true
            }
        };

        meet(&mut self.heap, other.heap)
            | meet(&mut self.table, other.table)
            | meet(&mut self.vmctx, other.vmctx)
            | meet(&mut self.other, other.other)
    }
```

cfallin replied:
The second version reads more confusingly to me; and it seems less error-prone to retain the top-level "if old != new" check than to thread a bool through the recomputation. If you really feel strongly here I'm happy to tweak further though...

jameysharp replied:
Nah, I don't feel strongly about this.

jameysharp reviewed (commented):
I'm sorry for this quantity of comments on codegen::egraph::stores specifically. A lot confused me here...

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+//! We can't say for sure that the "last store" *did* actually write
+//! that state, but we know for sure that no instruction *later* than
+//! it (up to the current instruction) did. However, we can get a
+//! must-alias property from this: if at a given load or store, we
+//! look backward to the "last store", *AND* we find that it has
+//! exactly the same address expression and type, then we know that
+//! the current instruction's access *must* be to the same memory
+//! location.
```
I think that updated version is actually subtly mismatching how the analysis works, and the version in the code is correct: the "last store" analysis doesn't consider store types, so we really are looking a the "last store" to the category, regardless of type or address. Then we compare address and type ("it has exactly the same address expression and type") and forward only if those match.

cfallin replied:
And specifically "an intervening store of a different type" would be caught by the logical condition described in the code, but not the modified text above.

jameysharp replied:
Maybe I'm confused by the use of the word "type". In this paragraph, I read that as referring to heap/table/vmctx/other. If you meant Cranelift type, like I8/I16/etc, that would make sense. Maybe saying "value type" disambiguates enough that I wouldn't have tripped over it?

cfallin replied:
Ah, yes, it is indeed a value type; I'll clarify the wording. Thanks!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+    Store(Inst),
+    BeforeInst(Inst),
+    AfterInst(Inst),
+    Top,
```
Right, it's not fully a meet-lattice (or a proper "meet" binary operator), because meet takes a location and generates a different "bottom" value for each location.

Or more properly, it is a family of meet semilattices, indexed on location; each location actually meets over a slightly different semilattice with its own private "bottom".

All that's a fancy and fairly useless way of saying that we want to not fall back to a generic "unknown, we can't conclude anything" state: we can actually still name the state (every program point has some symbolic state), we just don't have a better name for it or a proof that it is equal to any other state in the program.

So -- given that it's definitely not a meet-semilattice -- your suggestions are great actually and I've updated things (and added doc-comments). Thanks!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+
+                let opcode = func.dfg[inst].opcode();
+                if opcode == Opcode::Load {
+                    let mem_state = state.get_load_input_state(func, inst);
```
> I'd like a comment here on why this is currently limited to the Load opcode specifically, rather than any instruction with that format. I assume it's something about store-to-load forwarding requiring that the load and store access the same memory width, and also have the same Value type, so that the value passed to the store can be substituted for the load. (I imagine future work could include replacing widening loads with sign- or zero-extend instructions given a suitable narrow store.) But it's not obvious to me how this interacts with redundant-load elimination.

Yep, that's exactly it. I've added a comment to that effect.

The interaction with RLE is that the RLE pass performs an action only if this analysis provides a "memory state" for an instruction, so if we withhold that state, then the analysis is effectively saying "this is not an analyzable load".

> [removing redundant match, and `for_flags` update]

Awesome, I like this much better -- done!

(I think this is a Lens in the Haskell sense, or at least a Functor, but I'm gonna look away from the abyss before it swallows me here)

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+            self.vmctx = MemoryState::AfterInst(inst);
+            self.other = MemoryState::AfterInst(inst);
+        } else if opcode.can_store() {
+            if let Some(memflags) = func.dfg[inst].memflags() {
```
> [handle stack stores specially]

Yep, we definitely could do this. None of the benchmarks I had been developing with use them (actually, `cranelift-wasm` doesn't use them at all) so I didn't have an easy way to test or evaluate the effect of this; it can definitely be done as future work.

> [treat some instructions only as "other"]

Unfortunately I don't think this is correct in all circumstances: an instruction about which we don't know anything has to clobber all segments of the state, not just "other" (i.e., it could alias with any of the categories).

It's possible that reasoning about the specific instruction formats that exist could lead us to something more precise here, but I really don't want to go down that road, because it bakes in knowledge of the opcode set as static and unchanging, and thus sets up an action-at-a-distance timebomb for when we later add some other instruction kind that involves memory. I'd rather have fallback logic here that basically says "if we don't know, clobber everything" -- it's the safe and obviously-correct (for some value of "obvious") approach :-)

jameysharp replied:
Somehow I assumed cranelift-wasm used stack slots, but I guess there really is no reason for it to. Huh!

Well, I don't think it matters for this PR, but perhaps `has_memory_fence_semantics` should still be updated to include `Debugtrap`? It does seem like an instruction that shouldn't have memory accesses reordered across it.

cfallin replied:
Ah, yes, I agree -- updated (and actually added `op if op.can_trap()` as an additional match arm).

jameysharp commented on cranelift/codegen/src/loop_analysis.rs:
```diff
@@ -228,6 +283,28 @@ impl LoopAnalysis {
     }
+
+    fn assign_loop_levels(&mut self) {
+        let mut stack: SmallVec<[Loop; 8]> = smallvec![];
+        for lp in self.loops.keys() {
+            if self.loops[lp].level == LoopLevel::invalid() {
+                stack.push(lp);
+                while let Some(&lp) = stack.last() {
+                    if let Some(parent) = self.loops[lp].parent.into() {
+                        if self.loops[parent].level != LoopLevel::invalid() {
+                            self.loops[lp].level = self.loops[parent].level.inc();
+                            stack.pop();
+                        } else {
+                            stack.push(parent);
+                        }
+                    } else {
+                        self.loops[lp].level = LoopLevel::root().inc();
+                        stack.pop();
+                    }
+                }
+            }
+        }
```
I think the following implementation is correct. Asymptotically, it has the same time complexity, but its space complexity is O(1) instead of O(n). Whether it's faster or slower in practice is another question entirely, of course. My guess is it's about the same.

```suggestion
        for lp in self.loops.keys() {
            if self.loops[lp].level == LoopLevel::invalid() {
                let mut assigned = self.loops[lp].parent;
                let mut level = 0usize;
                while let Some(parent) = assigned.expand() {
                    level += 1;
                    if self.loops[parent].level != LoopLevel::invalid() {
                        level += self.loops[parent].level.0.into();
                        break;
                    }
                    assigned = self.loops[parent].parent;
                }
                let mut cur = PackedOption::from(lp);
                while cur != assigned {
                    self.loops[cur.unwrap()].level = level.into();
                    cur = self.loops[cur.unwrap()].parent;
                    level -= 1;
                }
            }
        }
```

In writing this, though, I realized that `LoopLevel::root()` can't appear in `loops[lp].level` after analysis is done, which means I think zero would be a better choice for an "invalid" value. This would all be a little simpler if a loop level were represented as `Option<NonZeroU8>`. Then you can even replace `LoopLevel::inc` with `NonZeroU8::saturating_add`, which apparently was just added in Rust 1.64.

cfallin replied:
Ah, I see, this is the same trick we saw in the famous single-predecessor-chain traversals :-) OK, yeah, avoiding the `SmallVec` might be worthwhile here. I agree that the shallow loop nests are unlikely to stress either algorithm (so e.g. pointer-chasing overhead of traversing twice is likely negligible) so on balance let's go with yours as it requires less storage.

jameysharp replied:
Yup, same trick! I might be over-eager to apply that trick now, but it's a pretty good trick. 😁

I hope my version works; I meant it to be illustrative, more than an exact solution. I typed it directly into the GitHub comment, so I didn't even know whether it would compile…

I was briefly tempted to suggest a different trick: temporarily overwrite the parent pointers during the traversal so they instead point to the child you need to return to, then put back the original parent pointers as you walk back up the stack. That gives you the stack traversal pattern in O(1) additional space. It's also too clever/confusing to justify using here, I think. 😅 I believe I first learned about it in the context of some garbage collector, where it makes more sense.

cfallin replied:
It almost worked! The last `level -= 1` would underflow before we take the backedge and hit a loop-exiting condition; reconfigured things a bit to break before that.

jameysharp replied:
I think that means I got the edge cases wrong a different way: the smallest level it should assign is 1, so decrementing it an extra time in the last loop iteration should never underflow.

Also, I see you changed my usize to a u8 so you could directly construct a LoopLevel. I'd imagined a From impl that would saturate when converting from a usize that was too large. But this is fine too, at least in debug builds where overflow panics. Loop depth comes from untrusted input so I think that probably should be checked in release builds too, right? Restricting to 255-deep nested loops seems reasonable to me, anyway.

jameysharp commented on cranelift/filetests/filetests/egraph/remat.clif:
```diff
@@ -0,0 +1,35 @@
+; check:   block2:
+; nextln:     v3 = iconst.i32 42
+; nextln:     v4 = iadd.i32 v0, v3
+; nextln:     return v4
```
This rematerialization test case seems like an odd example since `block2` is strictly worse on any machine that has... um... at least one register, right?

It's a nice example of re-associating and constant-folding, though.

cfallin replied:
Right, it's testing specifically that remat is preferred even when GVN would do the opposite naturally. It looks silly here where register pressure is low, but it happened to help significantly with SpiderMonkey, so...

(The case where it's most interesting is when there are a whole bunch of values like `x+1`, `x+2`, ... rematerialized inside a loop or otherwise deeper into a function, and when these all would otherwise be hoisted by LICM or GVN; then we have just one register `x` live instead of all of the values)

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+            }
+
+            for succ in cfg.succ_iter(block) {
+                let succ_first_inst = func.layout.block_insts(succ).into_iter().next().unwrap();
```
It's almost like these convenience functions exist to increase my convenience! I should know our APIs better; thanks :-)

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+
+    /// Get the state seen by a load, if any.
+    pub fn get_state_for_load(&self, inst: Inst) -> Option<MemoryState> {
+        self.load_mem_state.get(&inst).cloned()
```
Oh, neat, I didn't know about `.copied()`; new idiom noted!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+        cfg: &ControlFlowGraph,
+    ) -> SecondaryMap<Block, LastStores> {
+        let mut block_input = SecondaryMap::with_default(LastStores::default());
+        block_input.resize(func.dfg.num_blocks());
```
Needs a constructor that takes a default *and* capacity, with refactors now; I was about to complain that such a thing doesn't exist on `SecondaryMap` but I just went to add it instead!

jameysharp replied:
But you're passing the, um, default `default` to `with_default`. So the existing `with_capacity` is fine, right?

cfallin replied:
Oh, huh, `None` is indeed the default for an `Option`. I guess I could have guessed that but it didn't occur to me. Thanks!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+
+        let block_input = analysis.compute_block_input_states(func, cfg);
+        analysis.compute_load_last_stores(func, block_input);
+        analysis
```
Ah, that's much cleaner!

I'm not sure why I had `#[inline(never)]` here; perhaps that's from some performance tweaking but I should have left a comment why. In the absence of such a comment, I'll delete these (it seems like it shouldn't matter); and can always add them back if something blows up in final measurements.

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+            for inst in func.layout.block_insts(block) {
+                log::trace!(
+                    "alias analysis: scanning at inst{} with state {:?} ({:?})",
+                    inst.index(),
```
Fixed!

I also went ahead and converted these `log::trace!`'s to `trace!`'s, to ensure we use our custom-defined macro that conditionally compiles (I often forget this!).

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+            for inst in func.layout.block_insts(block) {
+                state.update(func, inst);
+                log::trace!("after inst{}: state is {:?}", inst.index(), state);
+            }
```
Done!

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+//! enter a block no longer dominated by that block, we clear the info
+//! to "unknown". (We could do a fixpoint analysis instead and resolve
+//! merges this way, but in practice when iterating over a
+//! structured-code CFG in RPO, our approach will do just as well.)
```
Ah, yes, good catch; updated.

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,270 @@
+    ) -> SecondaryMap<Block, LastStores> {
+        let mut block_input = SecondaryMap::with_default(LastStores::default());
+        block_input.resize(func.dfg.num_blocks());
+        let mut queue: SmallVec<[Block; 8]> = smallvec![];
```
Fixed (with `worklist`).

jameysharp commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,263 @@
+                let succ_first_inst = func.layout.first_inst(succ).unwrap();
+                let succ_state = &mut block_input[succ];
+                let old = succ_state.clone();
+                succ_state
+                    .get_or_insert_with(|| LastStores::default())
+                    .meet_from(&state, succ_first_inst);
```
I think this should be `.get_or_insert(state)`, because we want `meet_from` to leave the state unchanged if this is the first time we're setting it. Or just:

```suggestion
                if let Some(succ_state) = succ_state.as_mut() {
                    succ_state.meet_from(&state, succ_first_inst);
                } else {
                    *succ_state = state;
                }
```

cfallin replied:
Ah, right; I still had the `Top` default state in my mind and had figured that meeting with an initial `Top` was harmless; but here the default is `Entry` so we need to special-case the first set.

jameysharp reviewed (commented):
Have you measured the performance impact of this PR with egraphs _disabled_? I see that `ScopedHashMap` is used by `simple_gvn`, which I was recently noticing takes a non-trivial amount of time. It looks to me like the `ScopedHashMap` changes should make some operations faster, so it could be an improvement, but it would be nice to check that it's not a regression.

I don't think I can review the code-motion changes in the ISLE preludes, on either the Rust or `.isle` sides. I tried some basic `diff`s but it looks like things are in a different order as well as being moved to different files.

Ignoring those, I think I just have `egraph.rs` and `elaborate.rs` left, which I'll have to tackle tomorrow.

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+        None
+    }
+
+    fn at_loop_level(&mut self, eclass: Id) -> Option<(u8, Id)> {
```
No reason at all -- updated!

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+                                    let store_addr = store_args[1];
+                                    let store_addr = egraph.egraph.canonical_id(store_addr);
+                                    let load_addr = egraph.egraph.canonical_id(*load_addr);
+                                    if store_addr == load_addr {
```
Done!

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+}
+
+pub(crate) fn store_to_load<'a>(id: Id, egraph: &mut FuncEGraph<'a>) -> Id {
+    if let Some(load_key) = egraph.egraph.classes[id].get_node() {
```
The optimization entrypoints are invoked for every new eclass node created (i.e., every time an enode is added to an eclass), so we only need to (and in fact should only, to avoid redundancy) examine just the latest enode in the eclass. I'll add a comment regarding this!

> Is it necessary for correctness to add the Load node to the egraph if the store gets forwarded to the load? Hypothetically, could the caller pass load_op, load_ty, etc to this function, and entirely skip constructing the Node::Load if it succeeds? I'm not asking you to change that in this PR, I just want to know if there are invariants that would break if someone did that.

I think that's technically possible, yeah, although there's one important case it would pessimize (in compile time, not code quality): having the `Load` node in the egraph as well lets other later `Load`s (with the same memory state) deduplicate with it. If we did not add the `Load` node to the egraph at all, then later `Load`s would eventually still end up with the forwarded stored value if the original did, because `store_to_load` is deterministic; but we would redo the work each time to determine that.

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+                    return Some(types.as_slice(&self.egraph.node_ctx.types)[0]);
+                }
+                &Node::Load { ty, .. } => return Some(ty),
+                _ => {}
```
That's a great question! I think I simply did not hit these cases while experimenting with this extractor with rewrite rules, but they absolutely should exist.

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+            );
+        }
+    }
+    ctx.egraph.subsume_ids.clear();
```
Actually, thinking about this has made me realize why both this line (and the `subsume` bool also) don't work: reentrancy of the rewriting! From within the call to `simplify` we can create a new node and enter `optimize_eclass` recursively; I think both a clear-at-beginning and clear-at-end potentially prematurely clear `subsume` directives that should be seen by an outer invocation. The most precise memory-reuse approach is probably to clear it when exiting the top level (based on the reentrancy limiter / depth counter) but in order to make reasoning about this easier I think we can probably just let it live for the duration of the egraph (function) rewrite.

cfallin replied:
Ah, that's a good point; I had been imagining a bool flag on the egraph itself rather than the per-ISLE-invocation context. Happy to revisit this (and actually the whole reentrancy mechanism maybe) in followup work...

cfallin commented on cranelift/codegen/src/egraph/stores.rs:
```diff
@@ -0,0 +1,263 @@
+//! same "last store", and (ii) the same address expression, and (iii)
+//! the same opcode-and-offset, will deduplicate (the first will be
+//! computed, and the later ones will use the same value). Furthermore
+//! we have a rewrite rule (in opts/store_to_load.isle) that rewrites
```
Ah, I had missed that, thanks! (Yeah, I tried to write store-to-load forwarding as a rule at first, but it turned out to be too... bespoke?... to build a bunch of extractors and special support for side-effecting insts in the general rewrite framework to reason about loads and stores.)

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,309 @@
+                                        offset: store_offset,
+                                        ..
+                                    },
+                                ) if *load_offset == *store_offset => {
```
I golfed this a bit more, starting from your suggestions (thanks!); I combined the three conditionals (type, offset, address) into one as well. Much better now!

I'm still noodling over the idea of making this an ISLE rule instead; maybe eventually it will be better that way, when we have more infrastructure to support non-pure ops, but I think I want to approach that idea incrementally / as a followup.

cfallin commented:
> Have you measured the performance impact of this PR with egraphs disabled? I see that ScopedHashMap is used by simple_gvn, which I was recently noticing takes a non-trivial amount of time. It looks to me like the ScopedHashMap changes should make some operations faster, so it could be an improvement, but it would be nice to check that it's not a regression.

I did earlier on, and testing again just now with SpiderMonkey.wasm, I see, uh... a 6% speedup?

```
cfallin@min:~/work/wasmtime % RAYON_NUM_THREADS=1 hyperfine './wasmtime.main compile ../wasm-tests/spidermonkey.wasm'
Benchmark 1: ./wasmtime.main compile ../wasm-tests/spidermonkey.wasm
  Time (mean ± σ):      6.028 s ±  0.196 s    [User: 5.821 s, System: 0.157 s]
  Range (min … max):    5.883 s …  6.483 s    10 runs

cfallin@min:~/work/wasmtime % RAYON_NUM_THREADS=1 hyperfine './wasmtime.egraph compile ../wasm-tests/spidermonkey.wasm'
Benchmark 1: ./wasmtime.egraph compile ../wasm-tests/spidermonkey.wasm
  Time (mean ± σ):      5.761 s ±  0.030 s    [User: 5.613 s, System: 0.139 s]
  Range (min … max):    5.702 s …  5.808 s    10 runs
```

(That's measuring on my M1 Air laptop, pinned to 1 core but otherwise no special measures. I verified looking at the `context.rs` diff again that the list of passes being run is not any different. I suspect the difference could be at least partly due to the faster `ScopedHashMap` but I can look a bit deeper into this)

> I don't think I can review the code-motion changes in the ISLE preludes, on either the Rust or .isle sides. I tried some basic diffs but it looks like things are in a different order as well as being moved to different files.

Yeah, sorry, that's a bit of a mess now due to all the rebasing -- there was a good amount of manual cherrypicking of new definitions. I'll see if I can clean up the diff a bit (given that this is hopefully merging soon there won't be other work that'll invalidate this with more conflicts later!).

> Ignoring those, I think I just have egraph.rs and elaborate.rs left, which I'll have to tackle tomorrow.

Thanks (and uh, bon voyage and good luck; those are some of the most exciting bits of this PR)!

fitzgen reviewed (commented):
Looking great!

jameysharp commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -0,0 +1,207 @@
+(rule (simplify (sshr $I64 (ishl $I64 (uextend $I64 x @ (eclass_type $I32)) (iconst _ (simm32 32))) (iconst _ (simm32 32))))
+      (sextend $I64 x))
+
+;; TODO: strength reduction: mul/div to shifts
+;; TODO: div/rem by constants -> magic multiplications
```
I'm not Chris, but I think they're intended for a follow-up. Compiling with egraphs is disabled by default for now, so it doesn't have to have every optimization yet. They're probably necessary before egraphs can be the default though, I'd think.

cfallin replied:
Yep, I'd prefer to get this in-tree and then continue adding to the body of opts, rather than holding back even longer.

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,296 @@
+pub type IdArray = EntityList<Id>;
+#[allow(dead_code)]
+pub type Unit = ();
+pub type Range = (usize, usize);
```
I'm not sure; it's pre-existing though (this definition was copied from the ISLE definition preludes for the backends). That change could be made across all backends if we want to look into it after this PR, maybe?

cfallin commented on cranelift/codegen/src/inst_predicates.rs:
```diff
@@ -37,6 +39,7 @@ fn is_load_with_defined_trapping(opcode: Opcode, data: &InstructionData) -> bool
 
 /// Does the given instruction have any side-effect that would preclude it from being removed when
 /// its value is unused?
+#[inline(always)]
```
This one is actually load-bearing IIRC, from performance tuning!

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,296 @@
+    if egraph.rewrite_depth > REWRITE_LIMIT {
+        return id;
+    }
+    egraph.rewrite_depth += 1;
```
Done (the latter)!

cfallin commented:
> Yeah, sorry, that's a bit of a mess now due to all the rebasing -- there was a good amount of manual cherrypicking of new definitions. I'll see if I can clean up the diff a bit (given that this is hopefully merging soon there won't be other work that'll invalidate this with more conflicts later!).

I've gone over the diff to the ISLE-side and Rust-side definitions of the prelude and ensured that the common bits, which are new files from git's point of view, now have contents that are strictly in the same order and exactly match the removed lines from the old (lowering-only) prelude. (Latest commit does some reordering to make this true.)

(This involved a "diff-of-diffs" workflow and was a little ugly; and I wish git had better ways of showing code provenance for cross-file moves; but in the end at least I'm confident no definition was inadvertently missed or reverted to an older version.)

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -0,0 +1,411 @@
+                let types = types.freeze(&mut self.node_ctx.types);
+
+                let mem_state = self.alias_analysis.get_state_for_load(inst);
+                let is_readonly_load = func.dfg[inst].opcode() == Opcode::Load && {
+                    let memflags = func.dfg[inst].memflags().unwrap();
+                    memflags.readonly() && memflags.notrap()
+                };
```
> I'm guessing that, unlike in the alias analysis, here it really isn't important that the opcode be exactly Opcode::Load. Sign- or zero-extending loads and vector loads should also be okay to treat as pure if they're both readonly and notrap, right?

In theory yeah, that's absolutely true. However making that change is a little more than `op.can_load()` because we want exactly those loads that are "normal" (take a value from memory with no special semantics or other side-effects), and `.can_load()` is a much broader predicate that includes e.g. atomics and CAS; and there's no "normal load but maybe extending" predicate. So I wanted to avoid untested, uncertain complexity.

Happy to address this in a followup if we find cases where it's important (or just in general), though!

cfallin replied:
(The suggestion is great though, applying that; thanks!)

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -0,0 +1,411 @@
+                func.dfg.num_values(),
+                func.dfg.value_lists.capacity(),
+            ),
+            side_effects: SecondaryMap::with_default(0..0),
+            side_effect_ids: vec![],
+            store_nodes: FxHashMap::default(),
+            blockparams: SecondaryMap::with_default(0..0),
```
Indeed, that works!

cfallin commented on cranelift/codegen/src/opts.rs:
```diff
@@ -0,0 +1,296 @@
+        addr: load_addr,
+        mem_state: MemoryState::Store(store_inst),
+        ..
+    } = load_key.node::<NodeCtx>(&egraph.egraph.nodes)
```
Ah! I had fought with this for a bit when I first wrote it to try to get the type-inference to work; it never occurred to me to just parameterize directly on the node type. That indeed works. Thanks!

cfallin commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,383 @@
+}
+
+impl NodeCtx {
+    pub(crate) fn with_capacity(types: usize, args: usize) -> Self {
```
I like that; named it `with_capacity_for_dfg` to make its semantics clearer (just `for_dfg` might imply that it somehow translates the contents of the DFG or something like that).

cfallin commented on cranelift/codegen/src/egraph/node.rs:
```diff
@@ -0,0 +1,383 @@
+    pub args: ListPool<Id>,
+}
+
+impl std::default::Default for NodeCtx {
```
Yep, verified it's not needed; deleted!

jameysharp reviewed (commented):
Here's review from `FuncEGraph::build`...

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -0,0 +1,412 @@
+                    self.stats.store_map_insert += 1;
+                }
+
+                let id = match (side_effect, mem_state, id) {
+                    (_, Some(..), NewOrExisting::New(id)) => {
+                        // Loads: do store-to-load forwarding, and
+                        // otherwise add to side-effecting roots.
+                        let opt_id = crate::opts::store_to_load(id, self);
+                        trace!("store_to_load: {} -> {}", id, opt_id);
+                        if opt_id == id {
+                            self.side_effect_ids.push(id);
+                            self.stats.side_effect_nodes += 1;
+                        }
+                        opt_id
+                    }
+                    (true, _, id) => {
+                        let id = id.get();
+                        self.side_effect_ids.push(id);
+                        self.stats.side_effect_nodes += 1;
+                        id
+                    }
+                    (false, _, NewOrExisting::New(id)) => {
+                        // Apply all optimization rules immediately; the
+                        // aegraph (acyclic egraph) works best when we do
+                        // this so all uses pick up the eclass with all
+                        // possible enodes.
+                        crate::opts::optimize_eclass(id, self)
+                    }
+                    (false, _, NewOrExisting::Existing(id)) => id,
+                };
```
Unfortunately I think this will be a meaningful performance regression -- I remember specifically restricting store-to-load logic to run only on loads while tuning.

Happy to add comments here to clarify as needed -- any thoughts (while keeping the structure the same)?

cfallin replied:
OK, talked with @jameysharp offline about this one and I think we can get around the perf regression with another conditional; and the rest of this is actually nice and clean and I like it better. So I'll accept this then tweak it a bit.

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -0,0 +1,412 @@
+            for inst in func.layout.block_insts(block) {
+                let side_effect = has_side_effect(func, inst)
+                    || (func.dfg[inst].opcode().can_load()
+                        && !func.dfg[inst].memflags().unwrap().readonly())
```
Fixed!

cfallin commented on cranelift/codegen/src/egraph.rs:
```diff
@@ -0,0 +1,412 @@
+                        mem_state,
+                        srcloc,
+                    }
+                } else if side_effect {
```
I'm afraid I'm not completely following this suggestion, sorry. If I understand correctly, you're proposing a change, then going through a (fairly complex) case analysis to argue why it's correct? Or is the concern that the code as-written is incorrect? (I agree that the `can_store()` check is redundant and I've removed it, thanks!)

I think that at the point we are looking at the definition of predicates and their exhaustive cases, and effectively "inlining" that in our reasoning, we are doing something wrong here: we're writing more brittle code that has less modularity. E.g. I'd prefer not to rely on the intersection of specific load flags in two different files to reason about removing a condition here.

I think that as-written, the conditions are not too bad to understand from first principles: we want to handle the four categories of (i) readonly loads, (ii) all other loads, (iii) all side-effecting ops, and (iv) all pure ops. I like that there are boolean flags that indicate each of these (except the implicit `Some`-ness of `mem_state` for loads; I can add `is_load` if you'd prefer there). I don't like inlining e.g. a `|| can_load` condition in this if-ladder: it makes the reasoning a lot muddier to me.

It's possible I've missed the whole point of this comment though; please do let me know if so!

cfallin replied:
OK, talked with @jameysharp about this one too and it makes more sense now -- accepting suggestion. Thanks!

jameysharp commented on cranelift/codegen/src/egraph/elaborate.rs:
```diff
@@ -0,0 +1,589 @@
+        for &(id, ty) in block_params {
+            let value = self.func.dfg.append_block_param(block, ty);
+            trace!(" -> block param id {:?} value {:?}", id, value);
+            self.id_to_value.insert_if_absent(
```
Out of curiosity: since block params are unique, each such enode _should_ be absent at this point, right?

cfallin replied:
Yep, but the only API we have is `insert_if_absent`, not `insert_if_absent_and_assert_otherwise` :-)

jameysharp reviewed (commented):
I think this is everything I have on elaboration. Gonna go back and recheck a couple other things but I'm almost done!

cfallin commented on cranelift/codegen/src/egraph/elaborate.rs:
```diff
@@ -0,0 +1,589 @@
+        self.elab_stack.push(ElabStackEntry::Start { id });
+        self.process_elab_stack();
+        debug_assert_eq!(self.elab_result_stack.len(), 1);
+        self.elab_result_stack.pop().expect("Must have one result")
```
Ah, here we are actually returning the result, so unfortunately we can't do that...

cfallin replied:
Err, apparently I'm reading too fast because I see you noted this above and further noted it isn't used :-) Yes, actually, this is a nice cleanup, thanks!

cfallin commented on cranelift/codegen/src/egraph/elaborate.rs:
```diff
@@ -0,0 +1,589 @@
+            Node::Inst { srcloc, .. } | Node::Load { srcloc, .. } => *srcloc,
+            _ => RelSourceLoc::default(),
+        };
+        let is_term = instdata.opcode().is_branch() || instdata.opcode().is_return();
```
This is different than `is_terminator` I think as it needs to cover conditional branches as well.

The mechanism actually has a pretty interesting and subtle reason for existing, which I've added in a longer comment here -- I had forgotten as well until I rediscovered it via test failures :-)

jameysharp reviewed (approved):
Ship it! :tada:

cfallin commented:
Thanks so much for the detailed review -- it was really helpful!

So ends one long saga, and begins another (actually using this thing) :-)

jameysharp commented:
> > Have you measured the performance impact of this PR with egraphs disabled? I see that ScopedHashMap is used by simple_gvn, which I was recently noticing takes a non-trivial amount of time. It looks to me like the ScopedHashMap changes should make some operations faster, so it could be an improvement, but it would be nice to check that it's not a regression.
> 
> I did earlier on, and testing again just now with SpiderMonkey.wasm, I see, uh... a 6% speedup?

I was curious about this so I've just compared the current merge-base from main (d68ca3711) with this branch (ec8a1a932). I see a 1% improvement in compile time/CPU cycles/instructions retired on the bz2, pulldown-cmark, and spidermonkey benchmarks. Not 6% but definitely nice!
